File: backend/web/searxng/state.go
package searxng

import (
	"encoding/json"
	"os"
	"path/filepath"
)

type runState struct {
	Port int `json:"port"`
	PID  int `json:"pid"`
}

func (m *Manager) statePath() string {
	return filepath.Join(m.dataDir, "run.json")
}

func (m *Manager) writeState(port, pid int) error {
	data, err := json.Marshal(runState{Port: port, PID: pid})
	if err != nil {
		return err
	}
	return os.WriteFile(m.statePath(), data, 0o600)
}

func (m *Manager) readState() (port, pid int, ok bool) {
	data, err := os.ReadFile(m.statePath())
	if err != nil {
		return 0, 0, false
	}
	var st runState
	if err := json.Unmarshal(data, &st); err != nil {
		return 0, 0, false
	}
	return st.Port, st.PID, st.Port > 0 && st.PID > 0
}

File: backend/web/searxng/state.ts
import path from "node:path";
import fs from "node:fs";
import { Effect } from "effect";
import { searxng_error, type searxng_error as searxng_error_type } from "./error";

export type run_state = {
  port: number;
  pid: number;
};

const state_path = (data_dir: string): string => path.join(data_dir, "run.json");

export const write_state = (
  data_dir: string,
  port: number,
  pid: number,
): Effect.Effect<void, searxng_error_type> =>
  Effect.gen(function* () {
    const data = new TextEncoder().encode(JSON.stringify({ port, pid } as run_state));
    const p = state_path(data_dir);
    yield* Effect.try({
      try: () => fs.writeFileSync(p, data, { mode: 0o600 }),
      catch: (cause) => searxng_error("write run state", cause),
    });
  });

export const read_state = (data_dir: string): [number, number, boolean] => {
  try {
    const p = state_path(data_dir);
    const data = fs.readFileSync(p);
    const st = JSON.parse(data.toString("utf8")) as run_state;
    return [st.port, st.pid, st.port > 0 && st.pid > 0];
  } catch {
    return [0, 0, false];
  }
};

